src/common.cpp

```diff
@@ -1,10 +1,93 @@
 #include "common.hh"
+#include "zarr.hh"
 
 #include "platform.h"
 
 #include <cmath>
+#include <thread>
 
 namespace common = acquire::sink::zarr::common;
+common::ThreadPool::ThreadPool(size_t n_threads,
+                               std::function<void(const std::string&)> err)
+  : contexts_{ n_threads }
+  , jobs_mutex_{ std::make_shared<std::mutex>() }
+  , error_handler_{ err }
+{
+    if (n_threads == 0) {
+        throw std::runtime_error("Cannot create thread pool with 0 threads.");
+    }
+
+    for (auto& ctx : contexts_) {
+        ctx.should_stop = false;
+        ctx.thread = std::thread([this, ctx_ = &ctx] { thread_worker_(ctx_); });
+        ctx.mutex = jobs_mutex_;
+    }
+}
+
+common::ThreadPool::~ThreadPool() noexcept
+{
+    await_stop();
+}
+
+void
+common::ThreadPool::push_to_job_queue(JobT&& job)
+{
+    std::scoped_lock lock(*jobs_mutex_);
+    jobs_.push(std::move(job));
+
+    for (auto& ctx : contexts_) {
+        ctx.cv.notify_one();
+    }
+}
+
+void
+common::ThreadPool::await_stop() noexcept
+{
+    // spin down threads
+    for (auto& ctx : contexts_) {
+        ctx.should_stop = true;
+        ctx.cv.notify_one();
+
+        if (ctx.thread.joinable()) {
```

Comment:
Would we ever expect a non-joinable thread in this thread pool?

Reply:
Yes, because `await_stop` is called from `Zarr::stop`, which is called at least once during setup before acquisition actually happens.